validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java

```diff
@@ -38,39 +39,102 @@ public class BeaconProposerPreparer implements ValidatorTimingChannel {
   private static final Logger LOG = LogManager.getLogger();
 
   private final ValidatorApiChannel validatorApiChannel;
-  private final ValidatorIndexProvider validatorIndexProvider;
+  private Optional<ValidatorIndexProvider> validatorIndexProvider;
   private final ProposerConfigProvider proposerConfigProvider;
   private final Spec spec;
   private final Optional<Eth1Address> defaultFeeRecipient;
   private boolean firstCallDone = false;
 
-  public BeaconProposerPreparer(
+  private final Map<Integer, BeaconPreparableProposer> runtimeProposerMap =
+      new ConcurrentHashMap<>();
+  private final Map<Integer, BeaconPreparableProposer> configuredProposerMap =
+      new ConcurrentHashMap<>();
+
+  BeaconProposerPreparer(
       ValidatorApiChannel validatorApiChannel,
       ValidatorIndexProvider validatorIndexProvider,
       ProposerConfigProvider proposerConfigProvider,
       Optional<Eth1Address> defaultFeeRecipient,
       Spec spec) {
+    this(
+        validatorApiChannel,
+        Optional.of(validatorIndexProvider),
+        proposerConfigProvider,
+        defaultFeeRecipient,
+        spec);
+  }
+
+  public BeaconProposerPreparer(
+      ValidatorApiChannel validatorApiChannel,
+      Optional<ValidatorIndexProvider> validatorIndexProvider,
+      ProposerConfigProvider proposerConfigProvider,
+      Optional<Eth1Address> defaultFeeRecipient,
+      Spec spec) {
     this.validatorApiChannel = validatorApiChannel;
     this.validatorIndexProvider = validatorIndexProvider;
     this.proposerConfigProvider = proposerConfigProvider;
     this.defaultFeeRecipient = defaultFeeRecipient;
     this.spec = spec;
   }
 
+  public void initialize(final Optional<ValidatorIndexProvider> provider) {
```

Comment:
Feels odd to specify this in the constructor and here.  And possibly this is more honest to just call `setValidatorIndexProvider`?